# 積分

積分は2次元であればある区間の面積を、3次元であればとある領域の体積を求められます。
厳密には極限が必要なため、厳密な積分の実装はできません。しかし数学でも積分を習う前には **区分求積法** を習いますが、こちらであれば実装は可能です。

$y=x^2$と$y=0$の間の図形の面積（の近似）を、$x=0$から$x=3$の区間を区分求積法で求めてみましょう。

In [42]:
# 等分数
N = 10

# 面積
s = 0.0

function f(x)
    return x^2
end

for n in 0:N-1
    s += (3/N) * f(3n/N)
end

println(s)
    

7.695


7.695という値になりました。これは具体的に何をしたかを説明します。

以下の図のように横軸の$x=0$と$x=3$を$N$等分します。

<img src="./images/02/区分求積法01.png" alt="Drawing" style="width: 500px;"/>

$x=0$から$x=3$までの長さは$3$ですので、$N$等分すると各小区間の長さは$3/N$です。次に縦軸を見ていきます。

<img src="./images/02/区分求積法02.png" alt="Drawing" style="width: 500px;"/>

横軸を分割して点を左から番号をつけていくと、各点の$x$座標は$n * 3/N$と表現できます。$n$が$0$の時には$0$で＄n=N＄のときには$3$ですね。すると分割した点ごとに$y=f(x)=x^2$の点は$f(n*3/N)=(n*3/N)^2$となります。

整理すると$x$軸を$N$個に分割して、それらの点と対応する$y=x^2$上の点を求めました。そしてこれをヒストグラムのようにつないでみると10本の長方形が現れました。これらの面積を求めると、求めたい図形の面積の近似を求めることができます。
$y=x^2$は曲線のため面積を求めるのは困難ですが、長方形の集まりであれば小学校でも習う「横かける縦」で計算したものを足し合わせれば求められます。
横の長さは先ほど確認した通りすべてが$3/N$であり、$n$番目の縦の長さは$(n*3/N)^2$でした。これを実装したものが先ほどのコードです。

しかし図でみるとわかるとおり、明らかに$y=x^2$と長方形の集まりには隙間があり、長方形の集まりの方が小さくなりそうです。今度は先ほどの長方形の作り方を変えてみましょう。

<img src="./images/02/区分求積法03.png" alt="Drawing" style="width: 500px;"/>

先ほどは長方形の「左肩」に$y=x^2$上の点をとりましたが、今度は「右肩」でとってみます。すると今度は明らかに、$y=x^2$より長方形の集まりの方が面積が大きそうです。
これをコードで確認してみましょう。




In [43]:
# 分割数
N = 10

# 面積
S = 0.0

function f(x)
    return x^2
end

for n in 1:N
    S += (3/N) * f(3n/N)
end

println(s)
    

7.695


コード自体はfor文の範囲を修正しただけですが、先ほどよりも大きくなりました。ですが明らかにはみ出しているので、元の図形の面積よりは常に大きくなりそうです。」

### 等分数

先ほどは等分を$10$で試しましたが、この数を増やすとどうなるでしょうか。左肩のときの場合で試してみます。

In [44]:
using Printf

for N=0:100:1000
    s = 0.0
    for n=0:N-1
        s += (3/N) * f(3n/N)
    end
    
    @printf("等分数%dのときの長方形の集まりの面積は%0.6f\n", N, s)
end
        

等分数0のときの長方形の集まりの面積は0.000000
等分数100のときの長方形の集まりの面積は8.865450
等分数200のときの長方形の集まりの面積は8.932612
等分数300のときの長方形の集まりの面積は8.955050
等分数400のときの長方形の集まりの面積は8.966278
等分数500のときの長方形の集まりの面積は8.973018
等分数600のときの長方形の集まりの面積は8.977512
等分数700のときの長方形の集まりの面積は8.980723
等分数800のときの長方形の集まりの面積は8.983132
等分数900のときの長方形の集まりの面積は8.985006
等分数1000のときの長方形の集まりの面積は8.986504


左肩の長方形の場合は、分割数を増やせば増やすほど徐々に増えていきます。今度は右肩の長方形の場合で見てみましょう。

In [45]:
using Printf

for N=0:100:1000
    S = 0.0    
    for n=1:N
        S += (3/N) * f(3n/N)
    end
    
    @printf("分割数%dのときの長方形の集まりの面積は%0.6f\n", N, S)
end
        

分割数0のときの長方形の集まりの面積は0.000000
分割数100のときの長方形の集まりの面積は9.135450
分割数200のときの長方形の集まりの面積は9.067612
分割数300のときの長方形の集まりの面積は9.045050
分割数400のときの長方形の集まりの面積は9.033778
分割数500のときの長方形の集まりの面積は9.027018
分割数600のときの長方形の集まりの面積は9.022512
分割数700のときの長方形の集まりの面積は9.019295
分割数800のときの長方形の集まりの面積は9.016882
分割数900のときの長方形の集まりの面積は9.015006
分割数1000のときの長方形の集まりの面積は9.013504


右肩の長方形の場合は、徐々に小さくなっていっています。

今回の$y=x^2$の場合は、分割数を増やせば増やすほど両者はある値に近づいていき、分割数を無限に増やすと両者は一つの値に収束します。この値が求めたい図形の面積であり、これが積分です。

区分求積法と積分を定義をします。

#### 定義: 区分求積法と面積

　関数$f(x)$に対して、$x$軸上のある区間$[a, b]$を$N$　個に等分し、更に
 
$$s(N) = \sum_{n=0}^{N-1} f(n \frac{b-a}{N}) \frac{b-a}{N}, S(N) =  \sum_{n=1}^{N} f(n \frac{b-a}{N}) \frac{b-a}{N}$$

と定義する。$N\rightarrow \infty$ の操作によって$s（N）$と$S(N)$が一つの値に定まるとき、その値を面積とする。

次に積分ですが、今からお話する積分は**リーマン積分**と呼ばれています。というのも、積分にはリーマン積分以外にもあるためです。
またほとんど区分求積法とは同じような考え方ですが、リーマン積分は**x**軸を「等分」するのではなく、単に**分割**します。
等分では面積が計算ができない場合が存在するためです。しかし分割して長方形の面積の和を求め、その分割数を無限に増やすという発想は同じです。

#### 　定義: 積分（リーマン積分）


　関数$f(x)$に対して、$x$軸上のある区間$[a, b]$を
 
$$a = x_0 < x_1 < x_2 < \dots < x_{N-2} < x_{N-1} < x_{N} = b $$

と$N$個に分割する。このとき

$$s(N) = \sum_{n=0}^{N-1} f(x_n) x_{n+1-n} , S(N) =  \sum_{n=1}^{N} f(x_n) x_n - x_{n-1}$$

と定義する。$N\rightarrow \infty$ の操作によって$s（N）$と$S(N)$が一つの値に定まるときリーマン積分可能といい**

$$  \int_{a}^{b} f(x) dx \Bigg( =  \lim_{N \rightarrow \infty} s(N) = \lim_{N\rightarrow \infty} S(N) \Bigg) $$

と書く。